.travis.yml

```diff
@@ -1,80 +1,89 @@
-  general:
     - USAGEREPORTER_Duplicati_LEVEL=none
     - AUTOUPDATER_Duplicati_SKIP_UPDATE=1
-    - TMPDIR=$HOME/tmp
+    - BUILD_DIR=${TRAVIS_BUILD_DIR}/../.build
+    - TEST_DIR=${TRAVIS_BUILD_DIR}/../.test
+    - ROOT_DIR="$TRAVIS_BUILD_DIR"
+
+# matrix expansion within jobs would be good:
+# https://github.com/travis-ci/travis-ci/issues/8295
+
+# There is one cache per branch, so concurrent builds on different branches possible!
+# DO NOT SET ENVIRONMENT VARIABLES WITHIN JOBS, OR CACHES WILL NOT BE SHARED AMONGST JOBS!
+
 
 jobs:
   include:
-    - stage: Run unittests
-      env: CATEGORY=Border
-    - env: CATEGORY=Filter
-
-    - env: CATEGORY=BulkNormal
+    - stage: build
       cache:
         directories:
-          - $HOME/download/bulk
+          - $BUILD_DIR
+      script:
+        - ${ROOT_DIR}/pipeline/jobs/build_job.sh
 
-    - env: CATEGORY=BulkNoSize
+
+    - stage: tests
       cache:
         directories:
-          - $HOME/download/bulk
-
-    - env: CATEGORY=SVNData
+          - $HOME/download/
+          - $BUILD_DIR
+      script:
+        - ${ROOT_DIR}/pipeline/jobs/unittest_job.sh --testcategories BulkNormal --testdata data.zip
+    - stage: tests
       cache:
         directories:
-          - $HOME/download/svn
-
-    - env: CATEGORY=SVNDataLong
+          - $HOME/download/
+          - $BUILD_DIR
+      script:
+        - ${ROOT_DIR}/pipeline/jobs/unittest_job.sh --testcategories BulkNoSize --testdata data.zip
+    - stage: tests
       cache:
         directories:
-          - $HOME/download/svn
-
-    - env: CATEGORY=Targeted
-    - env: CATEGORY=Purge
-    - env: CATEGORY=Serialization
-    - env: CATEGORY=Utility
-    - env: CATEGORY=UriUtility
-    - env: CATEGORY=WebApi
-    - env: CATEGORY=IO
-    - env: CATEGORY=ImportExport
-    - env: CATEGORY=Disruption
+          - $HOME/download/
+          - $BUILD_DIR
+      script:
+        - ${ROOT_DIR}/pipeline/jobs/unittest_job.sh --testcategories SVNDataLong,SVNData --testdata DSMCBE.zip
+    - stage: tests
+      cache:
+        directories:
+          - $HOME/download/
+          - $BUILD_DIR
+      script:
+        - ${ROOT_DIR}/pipeline/jobs/unittest_job.sh --testcategories Border
+    - stage: tests
+      cache:
+        directories:
+          - $HOME/download/
+          - $BUILD_DIR
+      script:
+        - ${ROOT_DIR}/pipeline/jobs/unittest_job.sh --testctegories Filter,Targeted,Purge,Serialization,WebApi,Utility,UriUtility,IO,ImportExport,Disruption
```

Comment:
Typo here?  `testctegories`

Reply:
Thanks for checking, and in fact travis doesn't bail out on errors, I have to look into that.

Reply:
Rootcause was refactoring from yesterday such that a trap was not set correctly. It works now. You can see an example unit test fail as the experiment.

Reply:
Thanks.  Does a failure in one category prevent the subsequent categories from being tested?  For example, the failure was in `IO`, so the `ImportExport` and `Disruption` tests don't run?  It would be nice if all test failures could be determined in a single build, similar to if we ran all the tests in one worker.

Reply:
Yes it prevents it. Often it is like that in CI/CD systems, fail fast. I know what you mean, and we can change this by caching the return values and check the aggregate later on, but do we really want that?

Reply:
Right now we have a dedicated worker for each test, the upshot is you see what fails but it's very expensive computational-wise :) I think it is not needed that much, if you do a PR it is a quality control for others to see there are no errors. Best practice would be, before a PR is made, that all tests locally run fine and travis shows the proof of that. With the start.sh script, you can now run these tests locally detached from your dev environment.

Reply:
That sounds reasonable.  Moreover, AppVeyor is running all the tests, so that would provide some more complete feedback as well.

By the way, I think this is great work.  The way we were distributing tests in Travis was really slow and frustrating.

Want to revert the deliberate test failure and we can see the tests pass again?

I think we're trying to impose a moratorium on merges so that we can incorporate the fix in pull request #4010 and hopefully ship a beta.  But since this only affects the Travis builds, I think this can be merged as well.